spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/config/DataFlowClientAutoConfiguration.java

```diff
@@ -139,28 +148,124 @@ public StreamBuilder streamBuilder(DataFlowOperations dataFlowOperations){
 		}
+
+		@Bean
+		OAuth2AccessTokenRefreshingClientHttpRequestInterceptor oAuth2AccessTokenRefreshingClientHttpRequestInterceptor
+				(OAuth2AccessTokenResponseProvider renewAccessTokenResponseProvider) {
+			return new OAuth2AccessTokenRefreshingClientHttpRequestInterceptor(renewAccessTokenResponseProvider);
+		}
+
+		@Bean
+		OAuth2AccessTokenResponseProvider refreshAccessTokenResponseProvider(ClientRegistrationRepository clientRegistrations,
+																			 OAuth2AccessTokenResponseClient clientCredentialsTokenResponseClient) {
+			return new OAuth2AccessTokenResponseProvider(clientRegistrations, clientCredentialsTokenResponseClient);
+		}
+
+		static class OAuth2AccessTokenResponseProvider {
+			private final OAuth2ClientCredentialsGrantRequest grantRequest;
+			private final OAuth2AccessTokenResponseClient clientCredentialsTokenResponseClient;
+
+			OAuth2AccessTokenResponseProvider(ClientRegistrationRepository clientRegistrations,
+											  OAuth2AccessTokenResponseClient clientCredentialsTokenResponseClient) {
+				ClientRegistration clientRegistration = clientRegistrations.findByRegistrationId(DEFAULT_REGISTRATION_ID);
+				this.grantRequest = new OAuth2ClientCredentialsGrantRequest(clientRegistration);
+				this.clientCredentialsTokenResponseClient = clientCredentialsTokenResponseClient;
+			}
+
+			public OAuth2AccessTokenResponse oAuth2AccessTokenResponse() {
+				return clientCredentialsTokenResponseClient.getTokenResponse(grantRequest);
+			}
+		}
+
+		static class OAuth2AccessTokenRefreshingClientHttpRequestInterceptor implements ClientHttpRequestInterceptor {
+
+			private RetryTemplate retryTemplate = new RetryTemplate();
+
+			private final OAuth2AccessTokenResponseProvider refreshAccessTokenResponseProvider;
+
+			private OAuth2AccessTokenResponse oAuth2AccessTokenResponse;
+
+			OAuth2AccessTokenRefreshingClientHttpRequestInterceptor(OAuth2AccessTokenResponseProvider refreshAccessTokenResponseProvider) {
+				Assert.notNull(refreshAccessTokenResponseProvider, "'refreshAccessTokenResponseProvider' cannot be null");
+				this.refreshAccessTokenResponseProvider = refreshAccessTokenResponseProvider;
+				this.oAuth2AccessTokenResponse = this.refreshOAuth2AccessTokenResponse();
+			}
+
+			@Override
+			public ClientHttpResponse intercept(HttpRequest httpRequest, byte[] bytes, ClientHttpRequestExecution execution) throws IOException {
+				return retryTemplate.execute(retryContext -> {
+					httpRequest.getHeaders().add("Authorization",
+							OAuth2AccessToken.TokenType.BEARER.getValue() + " " + this.getTokenValue());
+					logger.debug("Token issued " + oAuth2AccessTokenResponse.getAccessToken().getIssuedAt() +
+							" expires: " + oAuth2AccessTokenResponse.getAccessToken().getExpiresAt());
+					ClientHttpResponse response = execution.execute(httpRequest, bytes);
+					logger.debug(response.getStatusCode());
+					if (response.getStatusCode().equals(HttpStatus.UNAUTHORIZED)) {
```

Comment:
This is a problematic as you may get unauthorized even with valid access token and then retry template, afaik does it things 3 times on default. With client credentials flow you only care about if token has expired when to choose refresh it.